hibernate-core/src/test/java/org/hibernate/orm/test/embeddable/EmbeddableElementCollectionUpdateTest.java

```diff
@@ -0,0 +1,160 @@
+import jakarta.persistence.ElementCollection;
+import jakarta.persistence.Embeddable;
+import jakarta.persistence.Entity;
+import jakarta.persistence.FetchType;
+import jakarta.persistence.Id;
+import jakarta.persistence.Table;
+import org.assertj.core.api.Assertions;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+@DomainModel(annotatedClasses = EmbeddableElementCollectionUpdateTest.Person.class)
+@SessionFactory
+@TestForIssue(jiraKey = "HHH-16297")
+public class EmbeddableElementCollectionUpdateTest {
+
+	private Person thePerson;
+
+	@BeforeAll
+	public void prepare(SessionFactoryScope scope) {
+		scope.inTransaction( session -> {
+			Set<Address> addresses = new HashSet<>( Arrays.asList(
+					new Address( "via milano", "Roma" ),
+					new Address( "via Cartesio", "Milano" )
+			) );
+			thePerson = new Person( 7242000, "Claude", addresses );
+			session.persist( thePerson );
+		} );
+	}
+
+	@Test
+	public void removeElementAndAddNewOne(SessionFactoryScope scope) {
+		scope.inTransaction( session -> {
+			Person foundPerson = session.find( Person.class, thePerson.getId() );
+			Assertions.assertThat( foundPerson ).isNotNull();
+			Set<Address> addresses = foundPerson.getAddresses();
+			addresses.remove( new Address( "via milano", "Roma" ) );
+			addresses.add( new Address( "via Pierce", "Milano" ) );
+			assertThat( addresses )
+					.containsExactlyInAnyOrder(
+							new Address( "via Pierce", "Milano" ),
+							new Address( "via Cartesio", "Milano" )
+					);
+		} );
+		scope.inTransaction( session -> {
+			Person person = session.find( Person.class, thePerson.getId() );
+			assertThat( person.getAddresses() )
+					.containsExactlyInAnyOrder(
+							new Address( "via Pierce", "Milano" ),
+							new Address( "via Cartesio", "Milano" )
+					);
+		} );
+	}
+
+	@Entity(name = "Person")
+	@Table(name = "Person")
+	static class Person {
+		@Id
+		private Integer id;
+		private String name;
+
+		@ElementCollection(fetch = FetchType.EAGER)
+		private Set<Address> addresses;
+
+		public Person() {
+		}
+
+		public Person(Integer id, String name, Collection<Address> addresses) {
+			this.id = id;
+			this.name = name;
+			this.addresses = new HashSet<>( addresses );
+		}
+
+		public Integer getId() {
+			return id;
+		}
+
+		public void setId(Integer id) {
+			this.id = id;
+		}
+
+		public String getName() {
+			return name;
+		}
+
+		public void setName(String name) {
+			this.name = name;
+		}
+
+		public Set<Address> getAddresses() {
+			return addresses;
+		}
+
+		public void setAddresses(Set<Address> addresses) {
+			this.addresses = addresses;
+		}
+
+		@Override
+		public boolean equals(Object o) {
+			if ( this == o ) {
+				return true;
+			}
+			if ( o == null || getClass() != o.getClass() ) {
+				return false;
+			}
+			Person person = (Person) o;
+			return Objects.equals( name, person.name );
+		}
+
+	}
+
+	@Embeddable
+	public static class Address {
+		@Basic(optional = false)
```

Comment:
Would be interesting to have a test case where one of the attributes is also optional, to see if the null checks are properly handled during deletion.

Reply:
I'll add the test